File: swift/Sources/sdnx/Stringify.swift
import Foundation
import Collections

// MARK: - Status Type

private struct Status {
    var indent: Int
    var result: String
    var ansi: Bool
    var indentText: String
}

public struct StringifyOptions {
    var ansi: Bool?
    var indent: String?

    public init(ansi: Bool?, indent: String?) {
        self.ansi = ansi
        self.indent = indent
    }
}

// MARK: - Main Stringify Function

/// Converts an object into a string containing Structured Data Notation.
public func stringify(_ obj: OrderedDictionary<String, Any>, options: StringifyOptions? = nil) -> String {
    var status = Status(indent: 0, result: "", ansi: options?.ansi ?? false, indentText: options?.indent ?? "\t")
    printValue(obj, &status)
    return status.result
}

// MARK: - Private Helper Functions

private func printValue(_ obj: Any, _ status: inout Status) {
    if let array = obj as? [Any] {
        printArray(array, &status)
    } else if let date = obj as? Date {
        let dateStr = formatDate(date)
        if status.ansi {
            status.result += "\u{001B}[35m\(dateStr)\u{001B}[0m"
        } else {
            status.result += dateStr
        }
    } else if let dict = obj as? OrderedDictionary<String, Any> {
        printDictionary(dict, &status)
    } else if let dict = obj as? [String: Any] {
        printDictionary(OrderedDictionary(uniqueKeysWithValues: dict.map { ($0.key, $0.value) }), &status)
    } else if let str = obj as? String {
        if status.ansi {
            status.result += "\u{001B}[32m\"\(str)\"\u{001B}[0m"
        } else {
            status.result += "\"\(str)\""
        }
    } else if let bool = obj as? Bool {
        if status.ansi {
            status.result += "\u{001B}[34m\(bool)\u{001B}[0m"
        } else {
            status.result += "\(bool)"
        }
    } else if let num = obj as? NSNumber {
        // Handle both integers and floats
        if status.ansi {
            status.result += "\u{001B}[33m\(num)\u{001B}[0m"
        } else {
            status.result += "\(num)"
        }
    } else if obj is NSNull || obj is Void {
        status.result += "null"
    } else {
        status.result += "\(obj)"
    }
}

private func printArray(_ array: [Any], _ status: inout Status) {
    status.result += "["
    status.indent += 1

    for i in array.indices {
        status.result += "\n"
        indent(&status)
        printValue(array[i], &status)
        if i < array.count - 1 {
            status.result += ","
        }
    }

    status.indent -= 1
    status.result += "\n"
    indent(&status)
    status.result += "]"
}

private func printDictionary(_ dict: OrderedDictionary<String, Any>, _ status: inout Status) {
    status.result += "{"
    status.indent += 1

    let keys = Array(dict.keys)
    for i in keys.indices {
        status.result += "\n"
        let key = keys[i]
        indent(&status)
        status.result += "\(key): "
        if let value = dict[key] {
            printValue(value, &status)
        }
        if i < keys.count - 1 {
            status.result += ","
        }
    }

    status.indent -= 1
    status.result += "\n"
    indent(&status)
    status.result += "}"
}

private func indent(_ status: inout Status) {
    status.result += String(repeating: status.indentText, count: status.indent)
}

private func formatDate(_ date: Date) -> String {
    let calendar = Calendar.current
    let components = calendar.dateComponents([.year, .month, .day, .hour, .minute, .second], from: date)

    let year = components.year ?? 0
    let month = String(format: "%02d", components.month ?? 0)
    let day = String(format: "%02d", components.day ?? 0)
    let hours = String(format: "%02d", components.hour ?? 0)
    let minutes = String(format: "%02d", components.minute ?? 0)
    let seconds = String(format: "%02d", components.second ?? 0)

    if hours == "00" && minutes == "00" && seconds == "00" {
        return "\(year)-\(month)-\(day)"
    } else {
        return "\(year)-\(month)-\(day)T\(hours):\(minutes)"
    }
}

File: dotnet/Sdnx.Core/Stringify.cs
using System;
using System.Collections.Generic;
using System.Globalization;

namespace Sdnx.Core
{
    public class StringifyOptions
    {
        public bool Ansi { get; set; } = false;
        public string Indent { get; set; } = "\t";
    }

    public static class Stringify
    {
        private class Status
        {
            public int Indent { get; set; }
            public string Result { get; set; } = string.Empty;
            public bool Ansi { get; set; }
            public string IndentText { get; set; }

            public Status(bool ansi, string indentText)
            {
                Ansi = ansi;
                IndentText = indentText;
            }
        }

        /// <summary>
        /// Converts an object into a string containing Structured Data Notation.
        /// </summary>
        public static string Convert(Dictionary<string, object?> obj, StringifyOptions? options = null)
        {
            var status = new Status(
                options?.Ansi ?? false,
                options?.Indent ?? "\t"
            );

            PrintValue(obj, status);

            return status.Result;
        }

        private static void PrintValue(object? obj, Status status)
        {
            if (obj == null)
            {
                status.Result += "null";
            }
            else if (obj is List<object?> list)
            {
                status.Result += "[";
                status.Indent += 1;

                for (int i = 0; i < list.Count; i++)
                {
                    status.Result += "\n";
                    Indent(status);
                    PrintValue(list[i], status);
                    if (i < list.Count - 1)
                    {
                        status.Result += ",";
                    }
                }

                status.Indent -= 1;
                status.Result += "\n";
                Indent(status);
                status.Result += "]";
            }
            else if (obj is DateTime date)
            {
                string dateStr = FormatDate(date);
                status.Result += status.Ansi ? $"\u001b[35m{dateStr}\u001b[0m" : dateStr;
            }
            else if (obj is Dictionary<string, object?> dict)
            {
                status.Result += "{";
                status.Indent += 1;

                var keys = new List<string>(dict.Keys);
                for (int i = 0; i < keys.Count; i++)
                {
                    status.Result += "\n";
                    string key = keys[i];
                    Indent(status);
                    status.Result += key + ": ";
                    PrintValue(dict[key], status);
                    if (i < keys.Count - 1)
                    {
                        status.Result += ",";
                    }
                }

                status.Indent -= 1;
                status.Result += "\n";
                Indent(status);
                status.Result += "}";
            }
            else if (obj is string str)
            {
                status.Result += status.Ansi ? $"\u001b[32m\"{str}\"\u001b[0m" : $"\"{str}\"";
            }
            else if (obj is double num)
            {
                status.Result += status.Ansi ? $"\u001b[33m{num}\u001b[0m" : num.ToString();
            }
            else if (obj is int intNum)
            {
                status.Result += status.Ansi ? $"\u001b[33m{intNum}\u001b[0m" : intNum.ToString();
            }
            else if (obj is bool boolean)
            {
                var boolStr = boolean.ToString().ToLowerInvariant();
                status.Result += status.Ansi ? $"\u001b[34m{boolStr}\u001b[0m" : boolStr;
            }
            else
            {
                status.Result += obj.ToString();
            }
        }

        private static void Indent(Status status)
        {
            status.Result += new string(status.IndentText[0], status.Indent * status.IndentText.Length);
        }

        private static string FormatDate(DateTime date)
        {
            string year = date.Year.ToString();
            string month = date.Month.ToString().PadLeft(2, '0');
            string day = date.Day.ToString().PadLeft(2, '0');
            string hours = date.Hour.ToString().PadLeft(2, '0');
            string minutes = date.Minute.ToString().PadLeft(2, '0');
            string seconds = date.Second.ToString().PadLeft(2, '0');

            if (hours == "00" && minutes == "00" && seconds == "00")
            {
                return $"{year}-{month}-{day}";
            }
            else
            {
                return $"{year}-{month}-{day}T{hours}:{minutes}";
            }
        }
    }
}
